wagtail/admin/templates/wagtailadmin/pages/listing/_button_with_dropdown.html

```diff
@@ -1,20 +1,13 @@
 {% load wagtailadmin_tags %}
-<div {{ self.attrs }} class="c-dropdown  {% if is_parent %}t-inverted{% else %}t-default{% endif %} {{ classes|join:' ' }}" data-dropdown>
-    <a href="javascript:void(0)" aria-label="{{ title }}" class="c-dropdown__button u-btn-current {{button_classes|join:' ' }}">
-        {{ label }}
-        <div data-dropdown-toggle class="o-icon c-dropdown__toggle c-dropdown__togle--icon">
-            {% icon name="arrow-down" %}{% icon name="arrow-up" %}
-        </div>
-    </a>
-    <div class="t-dark">
-        <ul class="c-dropdown__menu u-toggle  u-arrow u-arrow--tl u-background">
-            {% for button in buttons %}
-                <li class="c-dropdown__item ">
-                    <a href="{{ button.url }}" aria-label="{{ button.attrs.title }}" class="u-link is-live {{ button.classes|join:' ' }}">
-                        {{ button.label }}
-                    </a>
-                </li>
-            {% endfor %}
-        </ul>
-    </div>
-</div>
+{% fragment as toggle_classname %}{{ classes|join:' ' }} {{ button_classes|join:' ' }}{% endfragment %}
+
+{% dropdown attrs=self.attrs toggle_icon="arrow-down" toggle_label=label toggle_aria_label=title toggle_classname=toggle_classname %}
```

Comment:
Unfortunately, this broke the bulk action buttons that are inside the dropdown. To replicate the bug, change the number `4` the `if len(bulk_actions_list) > 4:` block in `wagtailadmin_tags.bulk_action_choices` to something smaller e.g. `2`. Then, try to use one of the bulk actions inside the dropdown menu, e.g. "Publish".

I looked into this and it seems to be caused by a race condition between the `bulk-actions.js`'s `addBulkActionListeners` and the initialisation of `tippy()` in `DropdownController.ts`.

The `addBulkActionListeners` function attaches click listeners to the buttons (links) to add the appropriate `id` query parameters of the selected items.

It seems our Tippy/Stimulus initialisation is executed before bulk actions. Tippy hides the content on initialisation by removing it from the DOM, so the elements inside the dropdown do not get picked up by `querySelectorAll` in the bulk actions' initialisation.

If the `tippy()` initialisation is delayed, e.g. with a `setTimeout()`, this bug doesn't show up. This is of course not ideal, and will cause a flash of opened dropdown menus on initial load.

I tried putting `bulk-actions`' entrypoint in the webpack config to be the first one, but that didn't solve it. Calling `addBulkActionListeners()` directly instead of making it a listener to `DOMContentLoaded` also fixes the issue, but it may cause other issues if the bulk-actions code relies on other code. 

https://github.com/wagtail/wagtail/blob/401afe7a8170747ad687801b4c661eeac40a98a7/client/src/entrypoints/admin/bulk-actions.js#L6

(That said, even the `window.headerSearch` that comes after the above line might be brittle as it's not even wrapped in a `DOMContentLoaded` event handler.)

Not sure what the best course of action is... it seems the bulk actions feature needs a big rework, both on the frontend and backend, as it has been a pain point for me in the past too 😔

It makes more sense to do the rewrite later after we do Stimulus refactoring for other code that it interacts with, though (e.g. header search).

Reply:
Arh, I remember looking at the bulk actions JS and somehow forgot to update it. Thank you for catching this.

For now I have introduced a new "Tippy shown" custom event, which bulk actions can use to register their event listeners. Not super elegant but does the trick. If we want to refactor this further I feel like a `<form>` might be a decent option here compared to individual links for all actions?